### ISSUE TITLE
[question] What is the disposition of deps_user_info and user_info_build in Conan 2.0

### ISSUE THREAD

**Issue by System-Arch** [conan-io/conan#12812]
### What is your question?

Some CCI recipes (e.g., boost) are using deps_user_info, which does not appear to be supported in Conan 2.0 and thus results in a runtime error:

> -------- Testing the package --------
> boost/1.79.0 (test package): Writing generators to /tmp/tmp.5IEKaS9oOn/boost/test_package/test_output/build/generators
> boost/1.79.0 (test package): Generator 'CMakeDeps' calling 'generate()'
> boost/1.79.0 (test package): Generator 'VirtualRunEnv' calling 'generate()'
> boost/1.79.0 (test package): Calling generate()
> ERROR: boost/1.79.0 (test package): Error in generate() method, line 47
>         tc.variables["WITH_STACKTRACE_ADDR2LINE"] = self.deps_user_info["boost"].stacktrace_addr2line_available
>         AttributeError: 'TestPackageConan' object has no attribute 'deps_user_info'
> ERROR: Conan-2.0 test failed for 'boost'

What is the recommended strategy for dealing with such? Thanks

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
``tool_requires`` can pass information to its consumers via ``conf_info`` (in ``user.<pkgname>.xxxx:yyyy`` conf items), and consumers can read it via ``self.conf``.

Regular requires do not propagate this information directly, but consumers can access dependencies information via ``self.dependencies["mydep"]``. If there is some convention of things to find in the package, then accessing the dependencies package folder (or cpp_info.libdirs, cpp_info.bindirs) should be enough.

In general passing very custom and specific information to consumers via either mechanism (legacy deps_user_info or modern ``conf``), seems an anti-pattern and should probably be highly reduced to some extreme cases. I see it occurring often in all the autotools/automake/binutils/etc recipes in ConanCenter

``automake`` recipe define this:

```
        self.user_info.compile = os.path.join(self._automake_libdir, "compile")
        self.user_info.ar_lib = os.path.join(self._automake_libdir, "ar-lib")
```

And also the ``automake`` package declares several env-vars:
```python
        self.env_info.ACLOCAL.append(aclocal)

        automake_datadir = tools.unix_path(self._datarootdir)
        self.output.info("Setting AUTOMAKE_DATADIR to {}".format(automake_datadir))
        self.env_info.AUTOMAKE_DATADIR = automake_datadir

        automake_libdir = tools.unix_path(self._automake_libdir)
        self.output.info("Setting AUTOMAKE_LIBDIR to {}".format(automake_libdir))
        self.env_info.AUTOMAKE_LIBDIR = automake_libdir

        automake_perllibdir = tools.unix_path(self._automake_libdir)
        self.output.info("Setting AUTOMAKE_PERLLIBDIR to {}".format(automake_perllibdir))
        self.env_info.AUTOMAKE_PERLLIBDIR = automake_perllibdir

        automake = tools.unix_path(os.path.join(self.package_folder, "bin", "automake" + bin_ext))
        self.output.info("Setting AUTOMAKE to {}".format(automake))
        self.env_info.AUTOMAKE = automake
```
But a normal ``automake`` installation in a system or in msys2 do not define these variables. Are they really necessary? Why?

So in short:
- quick approach: use ``conf_info`` for ``tool_requires``, most of cases in ConanCenter
- long term approach: try to simplify recipes, learning why those variables are necessary, and investigating aternatives
- For regular ``requires``, there is nothing built-in. Packages could define their own conventions, like creating a json file in the root of the package with some information that consumers can read. This should happen practically never.

**Comment by System-Arch** [comment]
Hi @memsharded,

Thanks for the detailed reply. Here's the approach I came up with for the boost test package that was using the self.deps_user_info construct:

ORIGINAL CODE:
>        tc.variables["WITH_STACKTRACE_ADDR2LINE"] = self.deps_user_info["boost"].stacktrace_addr2line_available

MODIFIED CODE (WITHOUT self.deps_user_info):
>        tc.variables["WITH_STACKTRACE_ADDR2LINE"] = "stacktrace_addr2line" in self.dependencies["boost"].cpp_info.components

I'm leveraging the existence (or absence) of the component directly without needing an extra piece of data. Does this look okay? Thanks

**Comment by SpaceIm** [comment]
> `automake` recipe define this:
> 
> ```
>         self.user_info.compile = os.path.join(self._automake_libdir, "compile")
>         self.user_info.ar_lib = os.path.join(self._automake_libdir, "ar-lib")
> ```

It's a convenient variable we can use in msvc build of autotools based recipes (so that downstream recipes don't have to hardcode relative path of these scripts located in automake package folder, it's done once in automake recipe). `compile` & `ar-lib` scripts allow to convert GNU options to cl/lib options, it's the canonical robust way to compile autotools project with msvc: `export CC=<path/to/compile> cl`, `export CXX=<path/to/compile> cl` & `export AR=<path/to/ar-lib> lib`.

> And also the `automake` package declares several env-vars:
> 
> ```python
>         self.env_info.ACLOCAL.append(aclocal)
> 
>         automake_datadir = tools.unix_path(self._datarootdir)
>         self.output.info("Setting AUTOMAKE_DATADIR to {}".format(automake_datadir))
>         self.env_info.AUTOMAKE_DATADIR = automake_datadir
> 
>         automake_libdir = tools.unix_path(self._automake_libdir)
>         self.output.info("Setting AUTOMAKE_LIBDIR to {}".format(automake_libdir))
>         self.env_info.AUTOMAKE_LIBDIR = automake_libdir
> 
>         automake_perllibdir = tools.unix_path(self._automake_libdir)
>         self.output.info("Setting AUTOMAKE_PERLLIBDIR to {}".format(automake_perllibdir))
>         self.env_info.AUTOMAKE_PERLLIBDIR = automake_perllibdir
> 
>         automake = tools.unix_path(os.path.join(self.package_folder, "bin", "automake" + bin_ext))
>         self.output.info("Setting AUTOMAKE to {}".format(automake))
>         self.env_info.AUTOMAKE = automake
> ```

Several of these env vars are custom env vars injected through patches of automake recipe to allow relocatable automake installation (otherwise automake hardcodes several absolute paths, which is not a problem for system installation, but a blocker for conan). Their name is a little bit misleading, they should have been prefixed with `CONAN_` to clearly state that they are not regular automake variables, but internal recipe tricks.

**Comment by System-Arch** [comment]
Many of these use cases have now been eliminated by @jcar87's efforts to make the underlying scripts location-independent by using relative paths based on the location of the scripts themselves.
